platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/services/DefaultIsolatedProjectEvaluationListenerProvider.kt

```diff
@@ -38,25 +47,44 @@ class DefaultIsolatedProjectEvaluationListenerProvider : IsolatedProjectEvaluati
         actions.add(action)
     }
 
-    override fun isolate(): ProjectEvaluationListener? = when {
+    override fun isolateFor(gradle: Gradle): ProjectEvaluationListener? = when {
         actions.isEmpty() -> null
-        else -> IsolatedProjectEvaluationListener(ImmutableList.copyOf(actions).also { actions.clear() })
+        else -> {
+            val isolate = isolate(actions, IsolateOwner.OwnerGradle(gradle))
+            clear()
```

Comment:
My question about concurrency-safety should have been asked here instead. Is this method guaranteed to be called from a single thread, no need to worry about `actions` being used concurrently, etc?

Reply:
Yes. That's meant to be called once, after settings has been fully evaluated, currently in the `projectsLoaded` callback.